### PR TITLE
HAL_Linux: fix RPI Check

### PR DESCRIPTION
**Pull request by HefnySco**
[PR#22841](https://github.com/ArduPilot/ardupilot/pull/22841) assumes that  int UtilRPI::_check_rpi_version_by_rev() returns values from 0 to 3 which is not the case.
The modification applied made ardupilot wrongly detect RPI and as mentioned [here](https://discuss.ardupilot.org/t/obal-a-simple-linux-based-raspberry-pi-ardupilot-fcb/75479/101?u=mhefny)

I just removed the unnecessary modification.

### CODE REVIEW

**Comment by rsaxvc** [comment]
>https://github.com/ArduPilot/ardupilot/pull/22841 assumes that int UtilRPI::_check_rpi_version_by_rev() returns values from 0 to 3 which is not the case.

It's possible I've made a mistake, but at least for my Pi4:

```
rsaxvc@raspberrypi:~/code/ardupilot/libraries/AP_HAL_Linux $ xxd -g 1 /proc/device-tree/system/linux,revision
00000000: 00 b0 31 15                                      ..1.
rsaxvc@raspberrypi:~/code/ardupilot/libraries/AP_HAL_Linux $ dmesg | grep Machine\ model
[    0.000000] OF: fdt: Machine model: Raspberry Pi 4 Model B Rev 1.5
```

The top nibble of the third byte[index 2], is 3, not 4.

I suspect, but don't have a way to test,, that only this check might need removed, if PiZero and original Pi share the same /proc/device-tree/system/linux,revision

```
            if (_rpi_version==0) {
                _rpi_version=1; //RPI-Zero
            }
```

From https://forums.raspberrypi.com/viewtopic.php?t=326055 for Pi3: 
```
pi@raspberrypi:~ $ tail -3 /proc/cpuinfo
Revision        : a020d3
Serial          : 00000000cc2d52de
Model           : Raspberry Pi 3 Model B Plus Rev 1.3
pi@raspberrypi:~ $ od -An -tx1 /proc/device-tree/system/linux,revision
 00 a0 20 d3
```

@HefnySco , do you have commit 53c1c235b890bba4b8 in your tree? I have it in mine, and I think _check_rpi_version_by_rev() still works this way, returning the top nibble from byte index 2 of /proc/device-tree/system/linux,revision

**Comment by HefnySco** [comment]
I wrote the original code -at least the latest version of it- and it was working fine for RPI zero, zero2, RPI4 & RPI CM4
and there was no need to change this. Your original problem was not being able to run it on RPI-4 because of [kernel problem](https://github.com/ArduPilot/ardupilot/pull/22841#issuecomment-1423598205) not because of a bug.
now we have a bug and this[ prohibit ppl.](https://discuss.ardupilot.org/t/obal-a-simple-linux-based-raspberry-pi-ardupilot-fcb/75479/101?u=mhefny) to run OBAL on RPI-Zero.
I think there is no need for the original fix as there were no bug at the beginning. However I kept your extra checks in the code -although I am not 100% believe it is necessary to add- but for other modification they introduced a bug.

**Comment by HefnySco** [comment]
the value (4)  is for CM4 not for RPI-4
pls check this [PR-19530](https://github.com/ArduPilot/ardupilot/pull/19530)  

> ```
> 
> ```
> 
> 
>     
>   
> 
> The top nibble of the third byte[index 2], is 3, not 4.
> 
> 
> I suspect, but don't have a way to test,, that only this check might need removed, if PiZero and original Pi share the same /proc/device-tree/system/linux,revision

Please check [PR-19530](https://github.com/ArduPilot/ardupilot/pull/19530)  what you refer to is a special fix for CM4 not RPI-4.

**Comment by rsaxvc** [comment]
Here's my Pi2:

```
rsaxvc@raspberrypi:~ $ dmesg | grep Machine\ model
[    0.000000] OF: fdt: Machine model: Raspberry Pi 2 Model B Rev 1.1
rsaxvc@raspberrypi:~ $ xxd -g 1 /proc/device-tree/system/linux,revision
00000000: 00 a2 10 41                                      ...A
```

I believe your patch is going to move Pi2 from BCM2709_PERIPHERAL_BASE to BCM2708_PERIPHERAL_BASE, which doesn't seem right.

I've put another patch up here: https://github.com/rsaxvc/ardupilot/tree/pizero_fix but won't be able to test it. From the description at https://discuss.ardupilot.org/t/obal-a-simple-linux-based-raspberry-pi-ardupilot-fcb/75479/98?u=mhefny it seems like it should though.

**Comment by rsaxvc** [comment]
>Please check https://github.com/ArduPilot/ardupilot/pull/19530 what you refer to is a special fix for CM4 not RPI-4. So the code will not longer recognize RPI-CM4. and in function like void GPIO_RPI::init() it will simply crash with CM4.

I don't believe this is used anymore. It looks like RPI-4 and CM4 would both read /proc/device-tree/system/linux,revision . Do you know what it would report for CM4? I suspect we'd want to treat it the same as RPI-4, and we would if the revision file matched.

**Comment by lucasdemarchi** [comment]
We may need to stop trying to convert to specific RPI versions and just do what we need based on the soc. That is similar to how the other boards are handled.

I know in past I suggested reading the device-tree, but I'm starting to think that moving all Linux boards to just read it from /proc/cpuinfo would be better

**Comment by HefnySco** [comment]
> > Please check #19530 what you refer to is a special fix for CM4 not RPI-4. So the code will not longer recognize RPI-CM4. and in function like void GPIO_RPI::init() it will simply crash with CM4.
> 
> I don't believe this is used anymore. It looks like RPI-4 and CM4 would both read /proc/device-tree/system/linux,revision . Do you know what it would report for CM4? I suspect we'd want to treat it the same as RPI-4, and we would if the revision file matched.

yes It seems that my `[commit](https://github.com/ArduPilot/ardupilot/commit/53c1c235b890bba4b8eded0a0144164c70d17be8)`  made RPI not working .... my fault I apologize for this.

Regarding RPI-Zero yes it can be fixed as mentioned  [https://discuss.ardupilot.org/t/obal-a-simple-linux-based-raspberry-pi-ardupilot-fcb/75479/98?u=mhefny](https://discuss.ardupilot.org/t/obal-a-simple-linux-based-raspberry-pi-ardupilot-fcb/75479/98?u=mhefny)

**Comment by rsaxvc** [comment]
If we need to get PiZero fixed shortly that's ok with me. I'll be away for several weeks soon, but can try again when I get back. That said...

> We may need to stop trying to convert to specific RPI versions and just do what we need based on the soc. That is similar to how the other boards are handled.

This is what I thought @HefnySco was trying to do in 53c1c235b890bba. The value from /proc/device-tree/system/linux,revision follows the soc, rather than the board name, just minus 1 compared to how the values used to work:

    RPI-1A: revision 0
    RPI-ZeroW: revision 0(same soc as 1A)
    RPI-3: revision 2(from https://forums.raspberrypi.com/viewtopic.php?t=326055)
    RPI-ZeroW-2: revision 2(same soc as Pi3)
    Pi4: revision 3
    CM4: revision 3

If we wanted these to follow the old scheme, I think we would just need to add 1 to everything, rather than only adding 1 to only RPi1/RPi-Zero.

**Comment by HefnySco** [comment]
> If we need to get PiZero fixed shortly that's ok with me. I'll be away for several weeks soon, but can try again when I get back. That said...
> 
> > We may need to stop trying to convert to specific RPI versions and just do what we need based on the soc. That is similar to how the other boards are handled.
> 
> This is what I thought @HefnySco was trying to do in [53c1c23](https://github.com/ArduPilot/ardupilot/commit/53c1c235b890bba4b8eded0a0144164c70d17be8). The value from /proc/device-tree/system/linux,revision follows the soc, rather than the board name, just minus 1 compared to how the values used to work:
> 
> ```
> RPI-1A: revision 0
> RPI-ZeroW: revision 0(same soc as 1A)
> RPI-3: revision 2(from https://forums.raspberrypi.com/viewtopic.php?t=326055)
> RPI-ZeroW-2: revision 2(same soc as Pi3)
> Pi4: revision 3
> CM4: revision 3
> ```
> 
> If we wanted these to follow the old scheme, I think we would just need to add 1 to everything, rather than only adding 1 to only RPi1/RPi-Zero.

I believe we can follow the new scheme as and just fix this rpi-zero.

**Comment by rmackay9** [comment]
We discussed this on the dev call (06-June-2023) and @tridge agreed that we should use /proc/cpuinfo (@lucasdemarchi's suggestion above).

**Comment by tridge** [comment]
looks like we can get it from tree:
https://forums.raspberrypi.com//viewtopic.php?t=244031
image

**Comment by tridge** [comment]
we should add a get_peripheral_base() function using above code snippet

**Comment by HefnySco** [comment]
**New Readings using /proc/device-tree/soc/ranges**

RPI-Zero:
pi@raspberrypi:~ $ hexdump -C /proc/device-tree/soc/ranges 
00000000  7e 00 00 00 20 00 00 00  02 00 00 00              |~... .......|
0000000c

peri base is 20000000

Linux raspberrypi 5.10.103+ #1529 Tue Mar 8 12:19:18 GMT 2022 armv6l GNU/Linux

======================================================================================================
RPI-ZeroW2
pi@raspberrypi:~ $ hexdump -C /proc/device-tree/soc/ranges 
00000000  7e 00 00 00 3f 00 00 00  01 00 00 00 40 00 00 00  |~...?.......@...|
00000010  40 00 00 00 00 00 10 00                           |@.......|
00000018

peri base is 3f000000


Linux raspberrypi 5.15.83-v8+ #3 SMP PREEMPT Sun Jan 8 18:36:18 EET 2023 aarch64 GNU/Linux

======================================================================================================
RPI-4:	
pi@raspberrypi:~ $ hexdump -C /proc/device-tree/soc/ranges 
00000000  7e 00 00 00 00 00 00 00  fe 00 00 00 01 80 00 00  |~...............|
00000010  7c 00 00 00 00 00 00 00  fc 00 00 00 02 00 00 00  ||...............|
00000020  40 00 00 00 00 00 00 00  ff 80 00 00 00 80 00 00  |@...............|
00000030

peri base is fe000000

Linux raspberrypi 5.15.84-v8+ #1613 SMP PREEMPT Thu Jan 5 12:03:08 GMT 2023 aarch64 GNU/Linux

======================================================================================================
RPI-4:
pi@raspberrypicompiler:~ $  hexdump -C /proc/device-tree/soc/ranges
00000000  7e 00 00 00 00 00 00 00  fe 00 00 00 01 80 00 00  |~...............|
00000010  7c 00 00 00 00 00 00 00  fc 00 00 00 02 00 00 00  ||...............|
00000020  40 00 00 00 00 00 00 00  ff 80 00 00 00 80 00 00  |@...............|
00000030

peri base is fe000000

Linux raspberrypicompiler 5.10.63-v7l+ #1496 SMP Wed Dec 1 15:58:56 GMT 2021 armv7l GNU/Linux

======================================================================================================
RPI-CM4
pi@raspberrypi:~ $ hexdump -C /proc/device-tree/soc/ranges 
00000000  7e 00 00 00 00 00 00 00  fe 00 00 00 01 80 00 00  |~...............|
00000010  7c 00 00 00 00 00 00 00  fc 00 00 00 02 00 00 00  ||...............|
00000020  40 00 00 00 00 00 00 00  ff 80 00 00 00 80 00 00  |@...............|
00000030

peri base is fe000000

Linux raspberrypi 5.10.63-v7l+ #1496 SMP Wed Dec 1 15:58:56 GMT 2021 armv7l GNU/Linux

======================================================================================================
======================================================================================================
OrangePI Zero 2
$ hexdump -C /proc/device-tree/soc/ranges 
00000000  00 00 00 00 00 00 00 00  00 00 00 00 40 00 00 00  |............@...|
00000010


Linux orangepizero2 5.16.17-sun50iw9 #3.0.6 SMP Tue Aug 9 13:51:16 CST 2022 aarch64 GNU/Linux


=======================================================================================================
Mangopi MQ-Quad MCore H616 Module 4xA53

$ hexdump -C /proc/device-tree/soc/ranges 
00000000  00 00 00 00 00 00 00 00  00 00 00 00 40 00 00 00  |............@...|
00000010

peri base is 0

Linux orangepizero2 5.16.17-sun50iw9 #3.0.8 SMP Mon Feb 6 10:04:02 CST 2023 aarch64 GNU/Linux

=======================================================================================================
Banana PI
hexdump: /proc/device-tree/soc/ranges: No such file or directory

Linux bpi-iot-ros-ai 4.4.55-BPI-M2Z-Kernel #1 SMP PREEMPT Sun Apr 26 21:01:50 HKT 2020 armv7l GNU/Linux

**Comment by HefnySco** [comment]
I have updated the function to use **"/proc/device-tree/soc/ranges"**
It can distinguish between [RPIZ,1,2,3,4,Z2] as well as AllWinner H616 for Mango & BananaPiZero2
RPI_ZERO & 1  have the same config
RPI_2,3,Zero2 have the same config
RPI-4 has its own config.
Return value is enumeration now for more readbility.

**Comment by HefnySco** [comment]
@rsaxvc  Would you test on your side?

File: libraries/AP_HAL_Linux/Util_RPI.cpp
Comment (rmackay9) [inline libraries/AP_HAL_Linux/Util_RPI.cpp:41]: No big deal but we wouldn't used "unsigned char"... maybe just char is enough?

**Comment by rmackay9** [comment]
Great to see this. I've added to the dev call for next Tuesday.  Hopefully we can get testing done by then.

**Comment by rsaxvc** [comment]
@HefnySco , this detects the correct peripheral address on RPi2 as well. I won't have time to test much more.

**Comment by HefnySco** [comment]
> @HefnySco , this detects the correct peripheral address on RPi2 as well.


RPI2,3,Zero2 have the same peri base-address. I tested it on RPIZeroW2 and I check online result for RPI3 and it was the same as one reported by Z2. I do not have RPI-2 but it should work.

If you find sometime please share result of executing ` hexdump -C /proc/device-tree/soc/ranges `on RPI-2

**Comment by rsaxvc** [comment]
Pi2:

```
rsaxvc@raspberrypi:~ $ dmesg | grep Machine\ model
[    0.000000] OF: fdt: Machine model: Raspberry Pi 2 Model B Rev 1.1
rsaxvc@raspberrypi:~ $ hexdump -C /proc/device-tree/soc/ranges
00000000  7e 00 00 00 3f 00 00 00  01 00 00 00 40 00 00 00  |~...?.......@...|
00000010  40 00 00 00 00 04 00 00                           |@.......|
00000018

```

**Comment by HefnySco** [comment]
> 

Thank you this is exactly as RPI-W2 as expected. I believe now we are confident to proceed.

**Review by tridge — DISMISSED** [review]
commit msgs should be HAL_Linux:
happy to merge when @Williangalvani has tested

**Comment by rmackay9** [comment]
I've added to the EU Dev Call so we don't forget about this.  If we're lucky @Williangalvani will have tested by then.  If not we will push to next Tuesday's dev call

File: libraries/AP_HAL_Linux/Util_RPI.cpp
Comment (rsaxvc) [reply libraries/AP_HAL_Linux/Util_RPI.cpp:52]: I believe this was some leftover code from trying to treat boards according to the board name(0/1/2/3/4, where we want pi0 and pi1 to be the same because they have the same soc) vs the soc ID or soc ranges(where remapping pi0 to pi1 is unnecessary since the soc ID or ranges would be the same for those boards already.

File: libraries/AP_HAL_Linux/Util_RPI.cpp
Comment (Williangalvani) [inline libraries/AP_HAL_Linux/Util_RPI.cpp:72]:
```suggestion
            printf("Perephiral base address is %x\n", base);
```

found a (4 actually) typo =]

**Review by Williangalvani — APPROVED** [review]
looks good. Navigator still works fine.

File: libraries/AP_HAL_Linux/Util_RPI.h
Comment (peterbarker) [inline libraries/AP_HAL_Linux/Util_RPI.h:7]:
```suggestion
enum class BoardType {
```

This should really be `enum class`  This is a lot of namespace-polution otherwise.

Follow-up-PR would be nice :-)

File: libraries/AP_HAL_Linux/Util_RPI.cpp
Comment (peterbarker) [inline libraries/AP_HAL_Linux/Util_RPI.cpp:37]: This is kind of missing the point of the method now!

```suggestion
void UtilRPI::detect_board_type() 
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
@HefnySco this was approved and marked for merge, but I think that rename is quite an important change.

I'll merge this anyway if you want, but it really would deserve an immediate follow-up PR to fix that problem.

Want to fix this before merge?

File: libraries/AP_HAL_Linux/Util_RPI.cpp
Comment (HefnySco) [reply libraries/AP_HAL_Linux/Util_RPI.cpp:37]: then I suggest to rename it to **_get_board_type_using_peripheral_base** because this is what it does .
there is a function called **detect_board_type** that returns LINUX_BOARD_TYPE

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**
